test/e2e/validation.go

```diff
@@ -22,199 +22,211 @@ import (
-			apiResource := metav1.APIResource{
-				Group:      "apps",
-				Version:    "v1",
-				Kind:       "DaemonSet",
-				Name:       "daemonsets",
-				Namespaced: true,
-			}
-			enableTypeDirective := enable.NewEnableTypeDirective()
-			validFtc = enable.GenerateTypeConfigForTarget(apiResource, enableTypeDirective).(*v1beta1.FederatedTypeConfig)
-		}
-		// Using TestNamespaceName() will ensure that for cluster-scoped
-		// deployments, a different namespace from the kubefed system
-		// namespace is used to make certain that the validation admission
-		// webhook works across all namespaces.
-		validFtc.Namespace = f.TestNamespaceName()
-	})
-
-	It(fmt.Sprintf("should fail when an invalid %s is created or updated", resourceName), func() {
-		// This test also implicitly tests the successful creation of a valid
-		// resource.
-		By(fmt.Sprintf("Creating an invalid %s", resourceName))
-		invalidFtc := validFtc.DeepCopyObject().(*v1beta1.FederatedTypeConfig)
-		invalidFtc.Spec.FederatedType.Group = ""
-		err := client.Create(context.TODO(), invalidFtc)
-		if err == nil {
-			f.Logger().Fatalf("Expected error creating invalid %s = %+v", resourceName, invalidFtc)
-		}
-
-		By(fmt.Sprintf("Creating a valid %s", resourceName))
-		validFtcCopy := validFtc.DeepCopyObject().(*v1beta1.FederatedTypeConfig)
-		err = client.Create(context.TODO(), validFtcCopy)
-		if err != nil {
-			f.Logger().Fatalf("Unexpected error creating valid %s = %+v, err: %v", resourceName, validFtcCopy, err)
-		} else {
-			framework.AddCleanupAction(func() {
-				err := client.Delete(context.TODO(), validFtcCopy, validFtcCopy.Namespace, validFtcCopy.Name)
-				if err != nil && !apierrors.IsNotFound(err) {
-					f.Logger().Errorf("Error deleting %s %s: %v", resourceName, validFtcCopy.Name, err)
-				}
-			})
-		}
-
-		By(fmt.Sprintf("Updating with an invalid %s", resourceName))
-		invalidFtc = validFtcCopy.DeepCopyObject().(*v1beta1.FederatedTypeConfig)
-		invalidFtc.Spec.FederatedType.Kind = ""
-		err = client.Update(context.TODO(), invalidFtc)
-		if err == nil {
-			f.Logger().Fatalf("Expected error updating invalid %s = %+v", resourceName, invalidFtc)
-		}
+		namespace = f.TestNamespaceName()
 	})
 
-	When("running with namespace scoped deployment", func() {
-		It(fmt.Sprintf("should succeed when an invalid %s is created outside the kubefed system namespace", resourceName), func() {
-			if !framework.TestContext.LimitedScope {
-				framework.Skipf("Cannot run validation admission webhook namespaced test in a cluster scoped deployment")
+	resourcesToValidate := []string{ftc.ResourceName, kfcluster.ResourceName}
+	for i := range resourcesToValidate {
+		resourceName := resourcesToValidate[i]
+		vrt := newValidationResourceTest(resourceName)
+		vrt.initialize()
+		It(fmt.Sprintf("for %s should fail when an invalid %s is created or updated", resourceName, resourceName), func() {
+			// This test also implicitly tests the successful creation of a valid
```

Comment:
Is this comment accurate?  The block starting on L69 appears to be an explicit check for successful creation of a valid resource.

(No action required) It may be worth minimizing the indentation of this test code by extracting to a separate function.

Reply:
The comment was to clarify the `It` description in that this `It` will also test creation as a result of having to test if an invalid update fails. I've removed since there's a `By` that describes the valid creation.